cpp/src/lists/concatenate_rows.cu

```diff
@@ -0,0 +1,438 @@
+
+  // For an input table of `n` columns, if after interleaving we have the list offsets are
+  // [ i_0, i_1, ..., i_n, i_n+1, ..., i_2n, ... ] then to concatenate them just modify the offsets
+  // to be [ i_0, i_n, i_2n, i_3n, ... ].
+  auto const d_interleaved_offsets = lists_column_view(interleaved_columns->view()).offsets_begin();
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(num_output_lists + 1),
+    d_offsets,
+    [d_interleaved_offsets,
+     num_cols     = input.num_columns(),
+     table_dv     = *table_dv_ptr,
+     d_validities = validities.begin(),
+     has_null_mask] __device__(auto const idx) {
+      if (has_null_mask) {
+        auto const any_valid = thrust::any_of(
+          thrust::seq, table_dv.begin(), table_dv.end(), [idx](auto const& list_col) {
+            return list_col.is_valid(idx);
+          });
+        d_validities[idx] = static_cast<int8_t>(any_valid);
+      }
+      return d_interleaved_offsets[idx * num_cols];
+    });
+
+  auto [null_mask, null_count] = [&] {
+    return has_null_mask
+             ? cudf::detail::valid_if(
+                 validities.begin(), validities.end(), thrust::identity<int8_t>{}, stream, mr)
+             : std::make_pair(rmm::device_buffer{}, size_type{0});
+  }();
+
+  // The child column containing list entries is taken from the `interleaved_columns` column.
+  auto interleaved_columns_content = interleaved_columns->release();
+
+  return make_lists_column(
+    num_output_lists,
+    std::move(list_offsets),
+    std::move(interleaved_columns_content.children[lists_column_view::child_column_index]),
+    null_count,
+    null_count > 0 ? std::move(null_mask) : rmm::device_buffer{},
+    stream,
+    mr);
+}
+
+/**
+ * @brief Generate list offsets and list validities for the output lists column from the table_view
+ * of the input lists columns.
+ *
+ * This function is called only when (has_null_mask == true and null_policy == NULLIFY_OUTPUT_ROW).
+ */
+std::pair<std::unique_ptr<column>, rmm::device_uvector<int8_t>>
+generate_list_offsets_and_validities(table_view const& input,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  auto const num_output_lists = input.num_rows();
+  auto const table_dv_ptr     = table_device_view::create(input);
+
+  // The output offsets column.
+  static_assert(sizeof(offset_type) == sizeof(int32_t));
+  static_assert(sizeof(size_type) == sizeof(int32_t));
+  auto list_offsets = make_numeric_column(
+    data_type{type_id::INT32}, num_output_lists + 1, mask_state::UNALLOCATED, stream, mr);
+  auto const d_offsets = list_offsets->mutable_view().template begin<offset_type>();
+
+  // The array of int8_t to store validities for list elements.
+  auto validities = rmm::device_uvector<int8_t>(num_output_lists, stream);
```

Comment:
I'm curious. Why is this an `int8_t` and not a `bool`?

Reply:
I have very bad obsession about the storage of `std::vector<bool>` thus I always try to avoid using any kind of bool array.

Reply:
Ah, I see your point about `std::vector<bool>`. As far as I know, `rmm::device_uvector<bool>` doesn't pull that kind of stunt. @harrism et al would know best, of course.

Reply:
Whether or not `rmm::device_uvector<bool>` is similar to `std::vector<bool>`, stay away from any kind of bool array. Who knows if some days we want to construct/transfer data between `std::vector<bool>` and a device memory?